<a href="https://colab.research.google.com/github/Hashwineey/DataScience-Project2_SalesDataAnalysis/blob/main/P146219_Project2_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📊 Sales Data Analysis

## **Dataset Introduction**
This project is based on a **sales transaction dataset** that contains detailed records of customer purchases, product details, and demographic information. The dataset is used to analyze sales performance, customer behavior, and profitability trends.

### **Dataset Features**
This analysis involves total of 9 datasets consisting multiple key attributes, such as:

- **Customer Information**:  
  - `CustomerKey`: Unique identifier for each customer.  
  - `Gender`: The gender of the customer (Male/Female).  
  - `AgeClass`: Categorized age groups of customers.  
  - `AnnualIncomeClass`: Categorized income levels of customers.  
  - `Occupation`: The type of job held by the customer.  

- **Product Details**:  
  - `ModelName`: Name of the product.  
  - `CategoryName`: Product category (e.g., Bikes, Clothing, Accessories).  
  - `ProductPrice`: The selling price of the product.  
  - `ProductProfit`: The profit earned from selling the product.  

- **Transaction Details**:  
  - `OrderDate`: The date of purchase.  
  - `TotalProfit`: The total profit generated from a particular sale.  
  - `CustomerKey`: Identifies repeat purchases by the same customer.  

### **Dataset Source**
The https://www.kaggle.com/datasets/ukveteran/adventure-worksom **[insert dataset source]**, which includes real or simulated sales transaction recorFor the purpose of this analysis, the datasets will beetail business. It has been preprocessed to remove inconsistencies and ensure quality analysis.

---

### **Objective of the Analysis**
The goal of this analysis is to examine sales trends, customer behavior, and profitability metrics to derive valuable insights. These insights can help in **identifying high-value customers, optimizing product sales, and improving overall business profitability**.

---

🔍 **Next Steps:**  
- Load and explore the dataset.
- Perform data preprocessing and cleaning.
- Conduct exploratory data analysis (EDA).
- Derive insights and visualize key findings.

---



### **Data Science Project 2 Part 1: Data Cleaning**

### Import libraries and load csv files

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Load libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import seaborn as sns

In [ ]:
# Load datasets
calendar = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Calendar.csv')
customer = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Customers.csv', encoding='latin1')
#encoding = 'latin1' because the file is not in UTF-8 encoding
prod_cat = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Product_Categories.csv')
prod_subcat = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Product_Subcategories.csv')
prod = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Products.csv')
returns = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Returns.csv')
sales_2015 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2015.csv')
sales_2016 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2016.csv')
sales_2017 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2017.csv')
territories = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Territories.csv')

In [ ]:
# Print datasets' dimensions

# Dictionary to hold datasets
datasets = {
    "calendar": calendar,
    "customer": customer,
    "prod_cat": prod_cat,
    "prod_subcat": prod_subcat,
    "prod": prod,
    "returns": returns,
    "sales_2015": sales_2015,
    "sales_2016": sales_2016,
    "sales_2017": sales_2017,
    "territories": territories
}

# Loop through datasets and print shapes
for name, df in datasets.items():
    print(f"{name}: {df.shape}")

calendar: (876, 1)
customer: (18148, 13)
prod_cat: (4, 2)
prod_subcat: (37, 3)
prod: (293, 11)
returns: (1809, 4)
sales_2015: (2630, 8)
sales_2016: (23935, 8)
sales_2017: (29481, 8)
territories: (10, 4)


## Preview of each dataset

In [ ]:
calendar.head()

,Date
0,01-01-2015
1,01-02-2015
2,01-03-2015
3,01-04-2015
4,01-05-2015


In [ ]:
customer.head()

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner
0,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y
1,11001,MR.,EUGENE,HUANG,14-05-1965,S,M,eugene10@adventure-works.com,"$60,000",3,Bachelors,Professional,N
2,11002,MR.,RUBEN,TORRES,12-08-1965,M,M,ruben35@adventure-works.com,"$60,000",3,Bachelors,Professional,Y
3,11003,MS.,CHRISTY,ZHU,15-02-1968,S,F,christy12@adventure-works.com,"$70,000",0,Bachelors,Professional,N
4,11004,MRS.,ELIZABETH,JOHNSON,08-08-1968,S,F,elizabeth5@adventure-works.com,"$80,000",5,Bachelors,Professional,Y


In [ ]:
prod_cat.head()

,ProductCategoryKey,CategoryName
0,1,Bikes
1,2,Components
2,3,Clothing
3,4,Accessories


In [ ]:
prod_subcat.head()

,ProductSubcategoryKey,SubcategoryName,ProductCategoryKey
0,1,Mountain Bikes,1
1,2,Road Bikes,1
2,3,Touring Bikes,1
3,4,Handlebars,2
4,5,Bottom Brackets,2


In [ ]:
prod.head()

,ProductKey,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice
0,214,31,HL-U509-R,"Sport-100 Helmet, Red",Sport-100,"Universal fit, well-vented, lightweight , snap...",Red,0,0,13.0863,34.9900
1,215,31,HL-U509,"Sport-100 Helmet, Black",Sport-100,"Universal fit, well-vented, lightweight , snap...",Black,0,0,12.0278,33.6442
2,218,23,SO-B909-M,"Mountain Bike Socks, M",Mountain Bike Socks,Combination of natural and synthetic fibers st...,White,M,U,3.3963,9.5000
3,219,23,SO-B909-L,"Mountain Bike Socks, L",Mountain Bike Socks,Combination of natural and synthetic fibers st...,White,L,U,3.3963,9.5000
4,220,31,HL-U509-B,"Sport-100 Helmet, Blue",Sport-100,"Universal fit, well-vented, lightweight , snap...",Blue,0,0,12.0278,33.6442


In [ ]:
returns.head()

,ReturnDate,TerritoryKey,ProductKey,ReturnQuantity
0,18-01-2015,9,312,1
1,18-01-2015,10,310,1
2,21-01-2015,8,346,1
3,22-01-2015,4,311,1
4,02-02-2015,6,312,1


In [ ]:
sales_2015.head()

,OrderDate,StockDate,OrderNumber,ProductKey,CustomerKey,TerritoryKey,OrderLineItem,OrderQuantity
0,01-01-2015,21-09-2001,SO45080,332,14657,1,1,1
1,01-01-2015,05-12-2001,SO45079,312,29255,4,1,1
2,01-01-2015,29-10-2001,SO45082,350,11455,9,1,1
3,01-01-2015,16-11-2001,SO45081,338,26782,6,1,1
4,02-01-2015,15-12-2001,SO45083,312,14947,10,1,1


In [ ]:
sales_2016.head()

,OrderDate,StockDate,OrderNumber,ProductKey,CustomerKey,TerritoryKey,OrderLineItem,OrderQuantity
0,01-01-2016,17-10-2002,SO48797,385,14335,1,1,1
1,01-01-2016,30-09-2002,SO48802,383,24923,9,1,1
2,01-01-2016,29-11-2002,SO48801,326,15493,1,1,1
3,01-01-2016,16-11-2002,SO48799,352,26708,4,1,1
4,01-01-2016,16-12-2002,SO48798,369,23332,9,1,1


In [ ]:
sales_2017.head()

,OrderDate,StockDate,OrderNumber,ProductKey,CustomerKey,TerritoryKey,OrderLineItem,OrderQuantity
0,01-01-2017,13-12-2003,SO61285,529,23791,1,2,2
1,01-01-2017,24-09-2003,SO61285,214,23791,1,3,1
2,01-01-2017,04-09-2003,SO61285,540,23791,1,1,1
3,01-01-2017,28-09-2003,SO61301,529,16747,1,2,2
4,01-01-2017,21-10-2003,SO61301,377,16747,1,1,1


In [ ]:
territories.head()

,SalesTerritoryKey,Region,Country,Continent
0,1,Northwest,United States,North America
1,2,Northeast,United States,North America
2,3,Central,United States,North America
3,4,Southwest,United States,North America
4,5,Southeast,United States,North America


## Data Preprosessing

### Merge all datasets

#### 1. Join product_category and product_subcategory dataframes on ProductCategoryKey.
##### Essentially, we want the ProductCategoryKey from product_category to map on product_subcategory. The shape of the df should be (37,4)

In [ ]:
# Merge on 'ProductCategoryKey'
merged_df1 = pd.merge(prod_cat, prod_subcat, on='ProductCategoryKey', how='inner')

In [ ]:
merged_df1.head()

,ProductCategoryKey,CategoryName,ProductSubcategoryKey,SubcategoryName
0,1,Bikes,1,Mountain Bikes
1,1,Bikes,2,Road Bikes
2,1,Bikes,3,Touring Bikes
3,2,Components,4,Handlebars
4,2,Components,5,Bottom Brackets


In [ ]:
merged_df1.shape

(37, 4)

#### 2. Merge merged_df1 with product df on ProductSubcategoryKey. The shape of the df should be (293,14) since we are adding 3 columns from merged_df1 to the product dataframe.

In [ ]:
# Merge on 'ProductSubcategoryKey'
merged_df2 = pd.merge(prod, merged_df1, on='ProductSubcategoryKey', how='inner')
merged_df2.shape

(293, 14)

In [ ]:
merged_df2.head()

,ProductKey,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName
0,214,31,HL-U509-R,"Sport-100 Helmet, Red",Sport-100,"Universal fit, well-vented, lightweight , snap...",Red,0,0,13.0863,34.9900,4,Accessories,Helmets
1,215,31,HL-U509,"Sport-100 Helmet, Black",Sport-100,"Universal fit, well-vented, lightweight , snap...",Black,0,0,12.0278,33.6442,4,Accessories,Helmets
2,218,23,SO-B909-M,"Mountain Bike Socks, M",Mountain Bike Socks,Combination of natural and synthetic fibers st...,White,M,U,3.3963,9.5000,3,Clothing,Socks
3,219,23,SO-B909-L,"Mountain Bike Socks, L",Mountain Bike Socks,Combination of natural and synthetic fibers st...,White,L,U,3.3963,9.5000,3,Clothing,Socks
4,220,31,HL-U509-B,"Sport-100 Helmet, Blue",Sport-100,"Universal fit, well-vented, lightweight , snap...",Blue,0,0,12.0278,33.6442,4,Accessories,Helmets


#### 3. Merge merged_df2 with returns on ProductKey. The shape should be (1809,17) since we are adding 3 columns from returns to merged_df2

In [ ]:
# Merge on 'ProductSubcategoryKey'
merged_df3 = pd.merge(merged_df2, returns, on='ProductKey', how='inner')
merged_df3.shape

(1809, 17)

#### 4. Add 'Sales_2015', 'Sales_2016' and 'Sales_2017' as separate columns in Customer dataset, based on the 'CustomerKey'

In [ ]:
# Add a Year column to each sales dataset
sales_2015['Year'] = 2015
sales_2016['Year'] = 2016
sales_2017['Year'] = 2017

In [ ]:
# Merge customer dataset with sales_2015 dataset. The shape should be (2630, 21)
cust_2015 = pd.merge(customer, sales_2015, on='CustomerKey', how='inner')

In [ ]:
cust_2015.shape

(2630, 21)

In [ ]:
# Merge customer dataset with sales_2016 dataset. The shape should be (23935, 21)
cust_2016 = pd.merge(customer, sales_2016, on='CustomerKey', how='inner')
cust_2016.shape

(23935, 21)

In [ ]:
# Merge customer dataset with sales_2017 dataset. The shape should be (29481, 21)
cust_2017 = pd.merge(customer, sales_2017, on='CustomerKey', how='inner')
cust_2017.shape

(29481, 21)

In [ ]:
# Merge datasets on 'CustomerKey'
merged_df = cust_2015.merge(cust_2016, how='outer')
merged_df4 = merged_df.merge(cust_2017, how='outer')
merged_df4.shape

(56046, 21)

In [ ]:
merged_df4.head(3)

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year
0,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016
1,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,10-08-2003,SO57418,541,9,2,2,2016
2,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,11-08-2003,SO57418,530,9,3,2,2016


#### 5. Merge merged_df4 and merged_df3 on the common columns

In [ ]:
# Get the common columns between merged_df4 and merged_df3
common_columns = set(merged_df4.columns).intersection(merged_df3.columns)

# Print the common columns
print(common_columns)

{'ProductKey', 'TerritoryKey'}


In [ ]:
# Perform the merge based on the common columns
merged_df5 = pd.merge(merged_df4, merged_df3, on=list(common_columns), how='inner')
merged_df5.head()

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity
0,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,26-09-2016,1
1,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,06-10-2016,1
2,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,03-12-2016,1
3,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,15-01-2017,1
4,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,08-04-2017,1


In [ ]:
merged_df5.shape

(424467, 36)

In [ ]:
# To verify that there is no duplicates
duplicates = merged_df5[merged_df5.duplicated()]

# Display the duplicates
print(duplicates)

Empty DataFrame
Columns: [CustomerKey, Prefix, FirstName, LastName, BirthDate, MaritalStatus, Gender, EmailAddress, AnnualIncome, TotalChildren, EducationLevel, Occupation, HomeOwner, OrderDate, StockDate, OrderNumber, ProductKey, TerritoryKey, OrderLineItem, OrderQuantity, Year, ProductSubcategoryKey, ProductSKU, ProductName, ModelName, ProductDescription, ProductColor, ProductSize, ProductStyle, ProductCost, ProductPrice, ProductCategoryKey, CategoryName, SubcategoryName, ReturnDate, ReturnQuantity]
Index: []


No duplicated rows

#### 6. Merge merged_df5 and territories dataset on TerritoryKey, the shape should be (424467, 39)

In [ ]:
# Rename SalesTerritoryKey to TerritoryKey in territories dataset
territories = territories.rename(columns={'SalesTerritoryKey': 'TerritoryKey'})
territories.head()

,TerritoryKey,Region,Country,Continent
0,1,Northwest,United States,North America
1,2,Northeast,United States,North America
2,3,Central,United States,North America
3,4,Southwest,United States,North America
4,5,Southeast,United States,North America


In [ ]:
df = pd.merge(merged_df5, territories, on='TerritoryKey', how='inner')
df.head(3)

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent
0,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,26-09-2016,1,Australia,Australia,Pacific
1,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,06-10-2016,1,Australia,Australia,Pacific
2,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,03-12-2016,1,Australia,Australia,Pacific


In [ ]:
df.shape

(424467, 39)

### Data Cleaning

Now that we had merged all of the datasets (excluded calendar df since its not neccesary), we first need to identify the types of the data we have

In [ ]:
pd.set_option('display.max_columns', None)
df.head(3)

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent
0,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,26-09-2016,1,Australia,Australia,Pacific
1,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,06-10-2016,1,Australia,Australia,Pacific
2,11000,MR.,JON,YANG,08-04-1966,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,SO57418,488,9,5,1,2016,21,SJ-0194-S,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,Short sleeve classic breathable jersey with su...,Yellow,S,U,41.5723,53.99,3,Clothing,Jerseys,03-12-2016,1,Australia,Australia,Pacific


In [ ]:
# Identify the type of the data
print(df.dtypes)

CustomerKey                int64
Prefix                    object
FirstName                 object
LastName                  object
BirthDate                 object
MaritalStatus             object
Gender                    object
EmailAddress              object
AnnualIncome              object
TotalChildren              int64
EducationLevel            object
Occupation                object
HomeOwner                 object
OrderDate                 object
StockDate                 object
OrderNumber               object
ProductKey                 int64
TerritoryKey               int64
OrderLineItem              int64
OrderQuantity              int64
Year                       int64
ProductSubcategoryKey      int64
ProductSKU                object
ProductName               object
ModelName                 object
ProductDescription        object
ProductColor              object
ProductSize               object
ProductStyle              object
ProductCost              float64
ProductPri

In [ ]:
df['CategoryName'].unique()

array(['Clothing', 'Accessories', 'Bikes'], dtype=object)

In [ ]:
df['SubcategoryName'].unique()

array(['Jerseys', 'Tires and Tubes', 'Helmets', 'Touring Bikes',
       'Fenders', 'Mountain Bikes', 'Road Bikes', 'Bottles and Cages',
       'Caps', 'Bike Stands', 'Gloves', 'Shorts', 'Socks', 'Bike Racks',
       'Hydration Packs', 'Cleaners', 'Vests'], dtype=object)

In [ ]:
df['ProductStyle'].unique()

array(['U', '0', 'W'], dtype=object)

#### 1. Manage null values (if any)

In [ ]:
df.isna().sum()

,0
CustomerKey,0
Prefix,2837
FirstName,0
LastName,0
BirthDate,0
MaritalStatus,0
Gender,2837
EmailAddress,0
AnnualIncome,0
TotalChildren,0


##### Handle 'Gender' null values

In [ ]:
# Display all null values of Prefix
gender_null = df[df['Prefix'].isna()]
gender_null.head(3)

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent
1208,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,NaN,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.98,4,Accessories,Fenders,13-07-2016,1,Australia,Australia,Pacific
1209,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,NaN,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.98,4,Accessories,Fenders,05-09-2016,1,Australia,Australia,Pacific
1210,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,NaN,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.98,4,Accessories,Fenders,17-10-2016,1,Australia,Australia,Pacific


In [ ]:
len(gender_null['FirstName'].unique())

113

Total of 113 customers don't have their gender record. However, we can compute them based on their First Names.

In [ ]:
gender_null['FirstName'].unique()

array(['ALEJANDRO', 'WENDY', 'ANGELA', 'ANNA', 'CESAR', 'KATHERINE',
       'NAOMI', 'KATIE', 'ALISHA', 'TODD', 'ADRIANA', 'DARREN', 'JASMINE',
       'ALEXA', 'ROBYN', 'BRANDON', 'RENEE', 'TERESA', 'CATHERINE',
       'NICOLAS', 'BENJAMIN', 'NINA', 'MINDY', 'ABBY', 'MICAH', 'RICARDO',
       'HAROLD', 'MIGUEL', 'ARTHUR', 'LEVI', 'ANDREA', 'RUTH', 'SHAWN',
       'NATHANIEL', 'AMANDA', 'NOAH', 'CEDRIC', 'MAKAYLA', 'LATOYA',
       'KEITH', 'MISTY', 'SHEILA', 'ALEXANDRIA', 'MADISON', 'DOUGLAS',
       'RICKY', 'JOSE', 'CHRISTIAN', 'RACHEL', 'JONATHAN', 'EBONY',
       'GLORIA', 'CANDICE', 'SHANNON', 'TAMMY', 'ORLANDO', 'FERNANDO',
       'JENNY', 'JACK', 'JASON', 'JÉSUS', 'LUIS', 'GERALD', 'EMILY',
       'SHEENA', 'DAISY', 'JACQUELINE', 'EMMA', 'SAVANNAH', 'CHRISTOPHER',
       'KRISTI', 'JEROME', 'MARIE', 'BARBARA', 'MARCUS', 'PEDRO',
       'CYNTHIA', 'DENNIS', 'JULIA', 'VANESSA', 'GAIL', 'ALEXANDRA',
       'THOMAS', 'EDDIE', 'BRETT', 'LESLIE', 'ZACHARY', 'CASEY', 'KENDRA',
       '

From the names above, we can list out all the male names.

In [ ]:
male = ['DARREN','ARTHUR','CHRISTOPHER','LUIS','GERALD','JÉSUS','NICOLAS','CHRISTIAN','TODD','JOSE','MIGUEL',
       'JACK','NATHANIEL','DOUGLAS','DEVIN','LOUIS','KEITH','BRANDON','FERNANDO','MICAH','CESAR','GAIL','REGINALD',
       'CRAIG','ZACHARY','DALTON','ALEJANDRO','JAMIE','LESLIE','RICARDO','SHAWN','BENJAMIN','LEVI','TERRY','JONATHAN',
       'MARCUS','HAROLD','IAN','JEROME','RANDALL','ORLANDO','EDWARD','ALEXANDER','THOMAS','DENNIS','RICKY','PEDRO',
       'EDDIE','BRETT','WYATT','BARRY','NOAH','ROBYN','JASON','NEIL','CEDRIC']

# Fill null values of gender based on the names listed

df.loc[df['Gender'].isnull(), 'Gender'] = df.loc[df['Gender'].isnull(), 'FirstName'].map(
    lambda x: 'M' if x in male else 'F'
    )

df.isnull().sum()

,0
CustomerKey,0
Prefix,2837
FirstName,0
LastName,0
BirthDate,0
MaritalStatus,0
Gender,0
EmailAddress,0
AnnualIncome,0
TotalChildren,0


In [ ]:
# Check if the values are correctly replaced in Gender column
df[df['Prefix'].isnull()]

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,OrderNumber,ProductKey,TerritoryKey,OrderLineItem,OrderQuantity,Year,ProductSubcategoryKey,ProductSKU,ProductName,ModelName,ProductDescription,ProductColor,ProductSize,ProductStyle,ProductCost,ProductPrice,ProductCategoryKey,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent
1208,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,13-07-2016,1,Australia,Australia,Pacific
1209,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,05-09-2016,1,Australia,Australia,Pacific
1210,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,17-10-2016,1,Australia,Australia,Pacific
1211,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,17-01-2017,1,Australia,Australia,Pacific
1212,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,20-01-2017,1,Australia,Australia,Pacific
1213,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,12-02-2017,1,Australia,Australia,Pacific
1214,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,08-06-2017,1,Australia,Australia,Pacific
1215,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,09-04-2003,SO51561,485,9,2,2,2016,30,FE-6654,Fender Set - Mountain,Fender Set - Mountain,Clip-on fenders fit most mountain bikes.,NaN,0,0,8.2205,21.9800,4,Accessories,Fenders,11-06-2017,1,Australia,Australia,Pacific
1216,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,18-05-2003,SO51561,356,9,1,1,2016,1,BK-M68S-46,"Mountain-200 Silver, 46",Mountain-200,Serious back-country riding. Perfect for all l...,Silver,46,U,1117.8559,2071.4196,1,Bikes,Mountain Bikes,04-07-2016,1,Australia,Australia,Pacific
1217,11025,NaN,ALEJANDRO,BECK,23-12-1945,M,M,alejandro45@adventure-works.com,"$10,000",2,Partial High School,Clerical,Y,24-07-2016,29-04-2003,SO51561,472,9,3,1,2016,25,VE-C304-M,"Classic Vest, M",Classic Vest,"Light-weight, wind-resistant, packs to fit int...",Blue,M,U,23.7490,63.5000,3,Clothing,Vests,07-02-2017,1,Australia,Australia,Pacific


We can ignore the null values in 'ProductColor' and 'Prefix' since the columns are not part of the analysis

#### 2. Drop Unused Columns

In [ ]:
# Drop unneccesary columns
# We want to exclude ['Prefix','FirstName','LastName','EmailAddress','OrderNumber','ProductKey',
#'OrderLineItem','ProductSubcategoryKey','ProductSKU','ProductDescription','ProductCategoryKey', 'ProductColor']

df = df.drop(['Prefix','FirstName','LastName','EmailAddress','OrderNumber','ProductKey',
              'OrderLineItem','ProductSubcategoryKey','ProductSKU', 'ProductDescription','ProductCategoryKey','ProductColor'],
             axis=1)
df.shape

(424467, 27)

#### 3. Get customers' age from their birth date. And then, bin the age into groups

In [ ]:
df['BirthDate'].head()

,BirthDate
0,08-04-1966
1,08-04-1966
2,08-04-1966
3,08-04-1966
4,08-04-1966


In [ ]:
from datetime import datetime

# Convert BirthDate to datetime
df['BirthDate'] = pd.to_datetime(df['BirthDate'], format='%d-%m-%Y', errors='coerce')

# Check the converted column
print(df['BirthDate'].head())

0   1966-04-08
1   1966-04-08
2   1966-04-08
3   1966-04-08
4   1966-04-08
Name: BirthDate, dtype: datetime64[ns]


In [ ]:
# Define today's date in 'now' variable
now = datetime.now()
now

datetime.datetime(2025, 1, 26, 12, 15, 37, 250963)

In [ ]:
# Calucate customers' age from 'BirthDate'
df['Age'] = (now.year - df['BirthDate'].dt.year) - ((now.month - df['BirthDate'].dt.month) < 0)
df['Age'].head()

,Age
0,58
1,58
2,58
3,58
4,58


In [ ]:
# We can drop BirthDate column now
df = df.drop(columns=['BirthDate'])
df.head(3)

,CustomerKey,MaritalStatus,Gender,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,TerritoryKey,OrderQuantity,Year,ProductName,ModelName,ProductSize,ProductStyle,ProductCost,ProductPrice,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent,Age
0,11000,M,M,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,9,1,2016,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,26-09-2016,1,Australia,Australia,Pacific,58
1,11000,M,M,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,9,1,2016,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,06-10-2016,1,Australia,Australia,Pacific,58
2,11000,M,M,"$90,000",2,Bachelors,Professional,Y,04-11-2016,06-09-2003,9,1,2016,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,03-12-2016,1,Australia,Australia,Pacific,58


In [ ]:
age = np.sort(df['Age'].unique())
print(age)

[ 44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61
  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79
  80  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97
  98  99 100 101 102 103 104 105 106 107 108 109 111 112 113 114]


In [ ]:
# Since we have customers whose age is more than 80, we will drop those customers as they are not our customer target group
# We need to know how many of them first
pd.set_option('display.max_rows', None)
count = df[df['Age'] > 80].shape[0]
print(count)

33650


In [ ]:
# Drop rows where Age > 80
df = df[df['Age'] <= 80]
df.shape

(390817, 27)

In [ ]:
# Get min and max of age
print("Max of Age: ", df['Age'].max())
print("Min of Age: ", df['Age'].min())
print("Median of Age: ", df['Age'].median())

Max of Age:  80
Min of Age:  44
Median of Age:  60.0


In [ ]:
# Define the age class based on age
bins = [40,50,60,70,80]
df['AgeClass'] = pd.cut(df['Age'], bins=bins)

<ipython-input-146-69710fc5f5bd>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['AgeClass'] = pd.cut(df['Age'], bins=bins)


In [ ]:
df['AgeClass'].value_counts()

,count
AgeClass,
"(50, 60]",140360
"(60, 70]",118739
"(70, 80]",69765
"(40, 50]",61953


#### 4. Convert all the date columns to datetime format

In [ ]:
# 'OrderDate'
df['OrderDate'] = pd.to_datetime(df['OrderDate'], format='%d-%m-%Y', errors='coerce')

# 'StockDate'
df['StockDate'] = pd.to_datetime(df['StockDate'], format='%d-%m-%Y', errors='coerce')

# 'Year'
df['Year'] = pd.to_datetime(df['Year'], format='%Y', errors='coerce')

# 'ReturnDate'
df['ReturnDate'] = pd.to_datetime(df['ReturnDate'], format='%d-%m-%Y', errors='coerce')

In [ ]:
# Check the format
print('OrderDate: ', df['OrderDate'].dtypes)
print('StockDate: ', df['StockDate'].dtypes)
print('Year: ', df['Year'].dtypes)
print('ReturnDate: ', df['ReturnDate'].dtypes)

OrderDate:  datetime64[ns]
StockDate:  datetime64[ns]
Year:  datetime64[ns]
ReturnDate:  datetime64[ns]


#### 5. Bin Annual Income

In [ ]:
df['AnnualIncome'].dtypes

dtype('O')

In [ ]:
# Remove dollar sign and commas, then convert to integer
df['AnnualIncome'] = df['AnnualIncome'].replace('[\$,]', '', regex=True).astype(int)

In [ ]:
df['AnnualIncome'].head()

,AnnualIncome
0,90000
1,90000
2,90000
3,90000
4,90000


In [ ]:
# Get min and max of AnnualIncome
print("Max of AnnualIncome: ", df['AnnualIncome'].max())
print("Min of AnnualIncome: ", df['AnnualIncome'].min())
print("Median of AnnualIncome: ", df['AnnualIncome'].median())

Max of AnnualIncome:  170000
Min of AnnualIncome:  10000
Median of AnnualIncome:  60000.0


In [ ]:
# Define the annual income class
bins = [0, 10000, 40000, 80000, 120000, 160000, 200000]
df['AnnualIncomeClass'] = pd.cut(df['AnnualIncome'], bins=bins)
df.head()

,CustomerKey,MaritalStatus,Gender,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,TerritoryKey,OrderQuantity,Year,ProductName,ModelName,ProductSize,ProductStyle,ProductCost,ProductPrice,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent,Age,AgeClass,AnnualIncomeClass
0,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-09-26,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]"
1,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-10-06,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]"
2,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-12-03,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]"
3,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2017-01-15,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]"
4,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2017-04-08,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]"


In [ ]:
df['AnnualIncomeClass'].value_counts()

,count
AnnualIncomeClass,
"(40000, 80000]",181710
"(10000, 40000]",120142
"(80000, 120000]",52131
"(0, 10000]",18289
"(120000, 160000]",16150
"(160000, 200000]",2395


#### 6. Calculate Product Profit from ProductCost and ProductPrice

In [ ]:
df['ProductProfit'] = (df['ProductPrice'] - df['ProductCost']).round(2)
df.head(3)

,CustomerKey,MaritalStatus,Gender,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner,OrderDate,StockDate,TerritoryKey,OrderQuantity,Year,ProductName,ModelName,ProductSize,ProductStyle,ProductCost,ProductPrice,CategoryName,SubcategoryName,ReturnDate,ReturnQuantity,Region,Country,Continent,Age,AgeClass,AnnualIncomeClass,ProductProfit
0,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-09-26,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]",12.42
1,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-10-06,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]",12.42
2,11000,M,M,90000,2,Bachelors,Professional,Y,2016-11-04,2003-09-06,9,1,2016-01-01,"Short-Sleeve Classic Jersey, S",Short-Sleeve Classic Jersey,S,U,41.5723,53.99,Clothing,Jerseys,2016-12-03,1,Australia,Australia,Pacific,58,"(50, 60]","(80000, 120000]",12.42


#### 7. Convert OrderDate from YYYY-MM-DD to YYYY-MM

In [ ]:
# Format OrderDate as YYYY-MM using strftime
df['OrderDate'] = df['OrderDate'].dt.strftime('%Y-%m')

# Display the updated DataFrame
print(df[['OrderDate']].head())

  OrderDate
0   2016-11
1   2016-11
2   2016-11
3   2016-11
4   2016-11


#### 8. Again, drop unnecessary columns

In [ ]:
df = df.drop(['StockDate','TerritoryKey','ReturnDate', 'TotalChildren','HomeOwner','ProductName','ProductSize','ProductStyle','Age'], axis=1)
df.head(3)

,CustomerKey,MaritalStatus,Gender,AnnualIncome,EducationLevel,Occupation,OrderDate,OrderQuantity,Year,ModelName,ProductCost,ProductPrice,CategoryName,SubcategoryName,ReturnQuantity,Region,Country,Continent,AgeClass,AnnualIncomeClass,ProductProfit
0,11000,M,M,90000,Bachelors,Professional,2016-11,1,2016-01-01,Short-Sleeve Classic Jersey,41.5723,53.99,Clothing,Jerseys,1,Australia,Australia,Pacific,"(50, 60]","(80000, 120000]",12.42
1,11000,M,M,90000,Bachelors,Professional,2016-11,1,2016-01-01,Short-Sleeve Classic Jersey,41.5723,53.99,Clothing,Jerseys,1,Australia,Australia,Pacific,"(50, 60]","(80000, 120000]",12.42
2,11000,M,M,90000,Bachelors,Professional,2016-11,1,2016-01-01,Short-Sleeve Classic Jersey,41.5723,53.99,Clothing,Jerseys,1,Australia,Australia,Pacific,"(50, 60]","(80000, 120000]",12.42


In [ ]:
df.isnull().sum()

,0
CustomerKey,0
MaritalStatus,0
Gender,0
AnnualIncome,0
EducationLevel,0
Occupation,0
OrderDate,0
OrderQuantity,0
Year,0
ModelName,0


#### 9. Drop duplicated rows

In [ ]:
# Find and drop duplicated rows

duplicates = df[df.duplicated()]
duplicates.shape

(339053, 21)

In [ ]:
# Drop duplicates in place
df.drop_duplicates(inplace=True)
df.shape

(51764, 21)

### Export cleaned dataset for data analysis

In [ ]:
# Save dataframe to CSV in Google Drive
path = '/content/drive/MyDrive/STQD6014_DataScience/Project 2/cleaned_data.csv'
df.to_csv(path, index=False)

# 🏢 Let's Recap!

## Dataset Overview
The dataset consists of sales transactions, customer demographics, and product details. After preprocessing, we retained the most relevant features for analysis. Below is a summary of the cleaned dataset:

### **🔹 Customer Information**
- **CustomerKey** – Unique identifier for each customer.
- **Gender** – Gender of the customer (`Male` / `Female`).
- **AgeClass** – Binned age group of customers (e.g., `(40, 50]`).
- **AnnualIncomeClass** – Categorized income levels (e.g., `(10000, 40000]`).
- **Occupation** – Customer’s job type (e.g., `Professional`, `Clerical`).

### **🔹 Product Details**
- **ModelName** – Name of the product purchased.
- **CategoryName** – Product category (e.g., `Bikes`, `Clothing`, `Accessories`).
- **ProductPrice** – Price at which the product was sold.
- **ProductProfit** – Profit earned from selling the product.

### **🔹 Transaction Details**
- **OrderDate** – Date of purchase.
- **TotalProfit** – The total profit generated from a sale.
- **RepeatCustomer** – Boolean (`True/False`) indicating if the customer has made multiple purchases.

These features were carefully selected after cleaning, merging, and removing redundant data. They serve as the foundation for our **customer analysis, sales trends, and profitability insights**.

---

## 🚀 Objective
The main goal of this analysis is to:
✅ Identify the most valuable customers.  
✅ Determine the most profitable products.  
✅ Analyze customer purchasing behavior.  
✅ Provide data-driven recommendations to improve profitability.  

---